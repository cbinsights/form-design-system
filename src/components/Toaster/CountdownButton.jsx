import React from 'react';
import PropTypes from 'prop-types';
import { Transition } from 'react-transition-group';
import FDS from 'lib/dictionary/js/styleConstants';
import IconButton from 'components/IconButton';
import DenyIcon from 'lib/icons/react/DenyIcon';

/**
 *
 * @param {Number} circleSize bounding box size of circle in px
 * @param {Number} strokeWidth width of circle stroke
 */
export const getCircleInfo = (circleSize, strokeWidth) => {
  const radius = (circleSize - strokeWidth) / 2;

  return {
    r: radius,
    c: Math.PI * radius * 2,
    centerOffset: circleSize / 2,
  };
};

const CountdownButton = ({ label = 'Close', Icon = DenyIcon, duration, onClick }) => {
  const circleSize = 34;
  const strokeWidth = '2';
  const { r, c, centerOffset } = getCircleInfo(circleSize, strokeWidth);

  const defaultCircleStyle = {
    // strokeDasharray sets up a dashed line in which each dash is equal to
    // the circumference of the circle.
    //
    // We offset the dash by 0 to start to create a full circle
    strokeDasharray: c,
    strokeDashoffset: 0,
    transition: `all ${duration}ms linear`,

    // start the countdown at "midnight" position of circle
    transformOrigin: 'center center',
    transform: 'rotate(-90deg)',
  };

  const transitionStyles = {
    entering: { strokeDashoffset: 0 },

    // offsetting the dash by the exact circle circumference lines up
    // the empty part of the dashed stroke with the circle circumference
    entered: { strokeDashoffset: c },
  };

  return (
    <div
      className="countdownButton"
      style={{
        width: `${circleSize}px`,
        height: `${circleSize}px`,
      }}
    >
      <div className="countdownButton-button">
        <IconButton onClick={onClick} radius="circle" Icon={Icon} label={label} />
      </div>
      <Transition
        in={true}
        appear
        timeout={{
          appear: 1,
          enter: duration,
        }}
        mountOnEnter
      >
        {(state) => (
          <svg
            className="countdownButton-animation"
            width={circleSize}
            height={circleSize}
            viewBox={`0 0 ${circleSize} ${circleSize}`}
          >
            <circle
              cx={centerOffset}
              cy={centerOffset}
              r={r}
              strokeWidth={strokeWidth}
              fill="transparent"
              stroke={FDS.FONT_COLOR_HINT}
              style={{
                ...defaultCircleStyle,
                ...transitionStyles[state],
              }}
            />
          </svg>
        )}
      </Transition>
    </div>
  );
};

CountdownButton.propTypes = {
  /** Duration in ms of countdown animation */
  duration: PropTypes.number.isRequired,

  /** Used to render a FDS Icon (should only be used for FDS Icons, e.g. Icon={DenyIcon}) */
  Icon: PropTypes.func,

  /** adds a label to IconButton */
  label: PropTypes.string,

  /** Click callback for button */
  onClick: PropTypes.func,
};

export default CountdownButton;
