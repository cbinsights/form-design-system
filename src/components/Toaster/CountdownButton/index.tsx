import React from 'react';
import { Transition } from 'react-transition-group';
import FDS from 'dictionary/js/styleConstants';
import IconButton from 'components/IconButton';
import DenyIcon from 'icons/react/DenyIcon';
import Icon from 'components/Icon';
export interface CircleInfo {
  radius: number;
  circumference: number;
  centerOffset: number;
}

/**
 *
 * @param {Number} circleSize bounding box size of circle in px
 * @param {Number} strokeWidth width of circle stroke
 */
export const getCircleInfo = (circleSize: number, strokeWidth: number): CircleInfo => {
  const radius = (circleSize - strokeWidth) / 2;

  return {
    radius,
    circumference: Math.PI * radius * 2,
    centerOffset: circleSize / 2,
  };
};

export interface CountdownButtonProps {
  /** Duration in ms of countdown animation */
  duration: number;

  /** Used to render a FDS Icon (should only be used for FDS Icons, e.g. Icon={DenyIcon}) */
  Icon?: Icon;

  /** adds a label to IconButton */
  label?: string;

  /** Click callback for button */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
}

const CountdownButton = ({
  label = 'Close',
  Icon = DenyIcon,
  duration,
  onClick,
}: CountdownButtonProps): JSX.Element => {
  const circleSize = 34;
  const strokeWidth = 2;
  const { radius, circumference, centerOffset } = getCircleInfo(circleSize, strokeWidth);

  const defaultCircleStyle = {
    // strokeDasharray sets up a dashed line in which each dash is equal to
    // the circumference of the circle.
    //
    // We offset the dash by 0 to start to create a full circle
    strokeDasharray: circumference,
    strokeDashoffset: 0,
    transition: `all ${duration}ms linear`,

    // start the countdown at "midnight" position of circle
    transformOrigin: 'center center',
    transform: 'rotate(-90deg)',
  };

  const transitionStyles = {
    entering: { strokeDashoffset: 0 },

    // offsetting the dash by the exact circle circumference lines up
    // the empty part of the dashed stroke with the circle circumference
    entered: { strokeDashoffset: circumference },
    exiting: undefined,
    exited: undefined,
    unmounted: undefined,
  };

  return (
    <div
      className="countdownButton"
      style={{
        width: `${circleSize}px`,
        height: `${circleSize}px`,
      }}
    >
      <div className="countdownButton-button">
        <IconButton onClick={onClick} radius="circle" Icon={Icon} label={label} />
      </div>
      <Transition
        in={true}
        appear
        timeout={{
          appear: 1,
          enter: duration,
        }}
        mountOnEnter
      >
        {(state) => (
          <svg
            className="countdownButton-animation"
            width={circleSize}
            height={circleSize}
            viewBox={`0 0 ${circleSize} ${circleSize}`}
          >
            <circle
              cx={centerOffset}
              cy={centerOffset}
              r={radius}
              strokeWidth={strokeWidth}
              fill="transparent"
              stroke={FDS.FONT_COLOR_HINT}
              style={{
                ...defaultCircleStyle,
                ...transitionStyles[state],
              }}
            />
          </svg>
        )}
      </Transition>
    </div>
  );
};

export default CountdownButton;
