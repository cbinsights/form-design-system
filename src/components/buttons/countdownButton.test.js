import React from 'react';
import { shallow, mount } from 'enzyme';

import CountdownButton, { getCircleInfo } from './CountdownButton';

const renderComponent = (props) => shallow(<CountdownButton {...props} />);

const MOCK_PROPS = { duration: 666 };

describe('CountdownButton', () => {

  it('matches snapshot', () => {
    const component = renderComponent(MOCK_PROPS);
    expect(component).toMatchSnapshot();
  });

  it('calculates circle dimensions correctly', () => {
    const actual = getCircleInfo(90, 10);
    const expected = {
      r: 40,
      c: Math.PI * 40 * 2,
      centerOffset: 45
    };

    Object.keys(expected).forEach((k) => {
      expect(actual[k]).toEqual(expected[k])
    })
  })

  describe('CountdownButton animation', () => {
    const component = mount(<CountdownButton {...MOCK_PROPS} />);
    const { strokeDashoffset } = component.find('circle').get(0).props.style;

    it('shows a full timer stroke on mount', () => {
      expect(strokeDashoffset).toBe(0);
    });

  });
});
