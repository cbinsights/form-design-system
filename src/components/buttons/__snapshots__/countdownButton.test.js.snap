// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountdownButton matches snapshot 1`] = `
<div
  className="countdownButton"
  style={
    Object {
      "height": "34px",
      "width": "34px",
    }
  }
>
  <div
    className="countdownButton-button"
  >
    <IconButton
      Icon={[Function]}
      radius="circle"
      size="m"
      theme="ghost"
    />
  </div>
  <Transition
    appear={true}
    enter={true}
    exit={true}
    in={true}
    mountOnEnter={true}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={
      Object {
        "appear": 1,
        "enter": 666,
      }
    }
    unmountOnExit={false}
  >
    <Component />
  </Transition>
</div>
`;
